File: protocol/src/pty-frame.ts
/**
 * Binary PTY output frame codec — the high-throughput half of the wire.
 *
 * Host -> controller terminal bytes travel as binary frames rather than JSON,
 * to avoid base64 bloat and per-frame parsing cost. Every frame is a fixed
 * 16-byte little-endian header followed by an opaque payload:
 *
 * ```
 * offset  size  field
 * 0       1     kind      (always 0x74, ASCII 't')
 * 1       1     version   (always 1)
 * 2       1     opcode    (PtyOpcode)
 * 3       1     reserved  (0)
 * 4       4     streamId  (u32 LE)
 * 8       8     seq       (u64 LE: low 32 bits, then high 32 bits)
 * 16      ...   payload
 * ```
 */

/** Frame magic: ASCII 't', marks a Pherry PTY frame. */
export const PTY_FRAME_KIND = 0x74

/** PTY framing version. Bumped only if the header layout changes. */
export const PTY_FRAME_VERSION = 1

/** Fixed header size in bytes. */
export const HEADER_BYTES = 16

/** What a PTY frame carries. */
export const PtyOpcode = {
  /** Incremental terminal output bytes. */
  Output: 1,
  /** Begin a full-screen snapshot. */
  SnapshotStart: 2,
  /** A chunk of the snapshot body. */
  SnapshotChunk: 3,
  /** End of the snapshot. */
  SnapshotEnd: 4,
  /** The PTY was resized. */
  Resized: 5,
  /** The session ended. */
  Ended: 6,
  /** A sequence gap was detected (frames were dropped). */
  Gap: 7,
} as const

/** One of the {@link PtyOpcode} values. */
export type PtyOpcode = (typeof PtyOpcode)[keyof typeof PtyOpcode]

/** A decoded PTY frame. */
export interface PtyFrame {
  opcode: PtyOpcode
  streamId: number
  seq: number
  payload: Uint8Array
}

/**
 * Encode a PTY frame to bytes.
 *
 * `seq` is a u64 split into two u32 halves — `seq >>> 0` (low) and
 * `Math.floor(seq / 2 ** 32)` (high) — so it survives JS's 53-bit-safe integer
 * range without BigInt. Both halves are written little-endian, low half first.
 */
export function encodePtyFrame(frame: PtyFrame): Uint8Array {
  const { opcode, streamId, seq, payload } = frame
  const out = new Uint8Array(HEADER_BYTES + payload.length)
  const view = new DataView(out.buffer)
  view.setUint8(0, PTY_FRAME_KIND)
  view.setUint8(1, PTY_FRAME_VERSION)
  view.setUint8(2, opcode)
  view.setUint8(3, 0)
  view.setUint32(4, streamId >>> 0, true)
  view.setUint32(8, seq >>> 0, true)
  view.setUint32(12, Math.floor(seq / 2 ** 32), true)
  out.set(payload, HEADER_BYTES)
  return out
}

/**
 * Decode a PTY frame, or return `null` if the buffer is too short or the magic
 * / version does not match (so callers can safely probe unknown bytes). Unknown
 * opcodes decode as-is for forward compatibility.
 */
export function decodePtyFrame(bytes: Uint8Array): PtyFrame | null {
  if (bytes.length < HEADER_BYTES) return null
  const view = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength)
  if (view.getUint8(0) !== PTY_FRAME_KIND) return null
  if (view.getUint8(1) !== PTY_FRAME_VERSION) return null
  const opcode = view.getUint8(2) as PtyOpcode
  const streamId = view.getUint32(4, true)
  const lo = view.getUint32(8, true)
  const hi = view.getUint32(12, true)
  const seq = hi * 2 ** 32 + lo
  const payload = bytes.slice(HEADER_BYTES)
  return { opcode, streamId, seq, payload }
}

File: ios/PherryKit/Sources/PherryKit/PtyFrame.swift
import Foundation

/// What a PTY frame carries.
public enum PtyOpcode: UInt8, Sendable {
    /// Incremental terminal output bytes.
    case output = 1
    /// Begin a full-screen snapshot.
    case snapshotStart = 2
    /// A chunk of the snapshot body.
    case snapshotChunk = 3
    /// End of the snapshot.
    case snapshotEnd = 4
    /// The PTY was resized.
    case resized = 5
    /// The session ended.
    case ended = 6
    /// A sequence gap was detected (frames were dropped).
    case gap = 7
}

/// Binary PTY output frame codec — the high-throughput half of the wire.
///
/// Host → controller terminal bytes travel as binary frames (not JSON) to avoid base64 bloat.
/// Every frame is a fixed 16-byte **little-endian** header followed by an opaque payload:
///
/// ```
/// offset  size  field
/// 0       1     kind      (always 0x74, ASCII 't')
/// 1       1     version   (always 1)
/// 2       1     opcode    (PtyOpcode)
/// 3       1     reserved  (0)
/// 4       4     streamId  (u32 LE)
/// 8       8     seq       (u64 LE: low 32 bits, then high 32 bits)
/// 16      ...   payload
/// ```
public struct PtyFrame: Sendable {
    /// Frame magic: ASCII 't'.
    static let kind: UInt8 = 0x74
    /// PTY framing version.
    static let version: UInt8 = 1
    /// Fixed header size in bytes.
    static let headerBytes = 16

    /// What the frame carries.
    public let opcode: PtyOpcode
    /// The per-connection PTY stream id (u32).
    public let streamId: UInt32
    /// The frame sequence number (u64).
    public let seq: UInt64
    /// The opcode-specific payload.
    public let payload: Data

    /// Build a PTY frame.
    public init(opcode: PtyOpcode, streamId: UInt32, seq: UInt64, payload: Data) {
        self.opcode = opcode
        self.streamId = streamId
        self.seq = seq
        self.payload = payload
    }

    /// Encode the frame to bytes (16-byte header || payload).
    public func encoded() -> Data {
        var out = Data(capacity: PtyFrame.headerBytes + payload.count)
        out.append(PtyFrame.kind)
        out.append(PtyFrame.version)
        out.append(opcode.rawValue)
        out.append(0)
        out.append(le32(streamId))
        out.append(le32(UInt32(seq & 0xffff_ffff)))
        out.append(le32(UInt32((seq >> 32) & 0xffff_ffff)))
        out.append(payload)
        return out
    }

    /// Decode a frame, or `nil` if the buffer is too short or the magic / version does not
    /// match — so a caller can safely probe unknown bytes. An unrecognised opcode byte also
    /// yields `nil`, since ``PtyOpcode`` models exactly the seven defined opcodes (the reference
    /// codec keeps unknown opcodes for forward compatibility; the controller ignores them, so
    /// the observable behaviour is unchanged).
    public static func decode(_ bytes: Data) -> PtyFrame? {
        guard bytes.count >= headerBytes else { return nil }
        let base = bytes.startIndex
        guard bytes[base] == kind, bytes[base + 1] == version else { return nil }
        guard let opcode = PtyOpcode(rawValue: bytes[base + 2]) else { return nil }
        let streamId = readLE32(bytes, at: 4)
        let low = UInt64(readLE32(bytes, at: 8))
        let high = UInt64(readLE32(bytes, at: 12))
        let seq = (high << 32) | low
        let payload = Data(bytes[(base + headerBytes)...])
        return PtyFrame(opcode: opcode, streamId: streamId, seq: seq, payload: payload)
    }

    /// Encode a `UInt32` as 4 little-endian bytes.
    private func le32(_ value: UInt32) -> Data {
        Data([
            UInt8(value & 0xff),
            UInt8((value >> 8) & 0xff),
            UInt8((value >> 16) & 0xff),
            UInt8((value >> 24) & 0xff),
        ])
    }

    /// Read a little-endian `UInt32` from `data` at `offset`.
    private static func readLE32(_ data: Data, at offset: Int) -> UInt32 {
        let i = data.startIndex + offset
        return UInt32(data[i])
            | (UInt32(data[i + 1]) << 8)
            | (UInt32(data[i + 2]) << 16)
            | (UInt32(data[i + 3]) << 24)
    }
}

/// Structured payloads for the non-`output` PTY frames — the codecs shared by encoder and
/// decoder. `snapshotStart` / `resized` carry a terminal size (`cols`, `rows` as u16 LE);
/// `ended` carries an exit code (an int32 LE, or an **empty** payload when the process was
/// killed by a signal → `nil`).
enum PtyPayload {
    /// Decode a size payload back to `(cols, rows)`.
    static func decodeSize(_ bytes: Data) -> (cols: Int, rows: Int) {
        guard bytes.count >= 4 else { return (0, 0) }
        let base = bytes.startIndex
        let cols = Int(bytes[base]) | (Int(bytes[base + 1]) << 8)
        let rows = Int(bytes[base + 2]) | (Int(bytes[base + 3]) << 8)
        return (cols, rows)
    }

    /// Encode a size payload: `cols`, `rows` as u16 LE.
    static func encodeSize(cols: Int, rows: Int) -> Data {
        Data([
            UInt8(cols & 0xff), UInt8((cols >> 8) & 0xff),
            UInt8(rows & 0xff), UInt8((rows >> 8) & 0xff),
        ])
    }

    /// Decode an exit-code payload: empty → `nil` (signalled), else the int32 LE code.
    static func decodeExit(_ bytes: Data) -> Int? {
        guard bytes.count >= 4 else { return nil }
        let base = bytes.startIndex
        let raw = UInt32(bytes[base])
            | (UInt32(bytes[base + 1]) << 8)
            | (UInt32(bytes[base + 2]) << 16)
            | (UInt32(bytes[base + 3]) << 24)
        return Int(Int32(bitPattern: raw))
    }
}
